boot/init/lib/mounting.rb

```diff
@@ -0,0 +1,77 @@
+module Mounting
+  def self.auto_depend_mount_points(mount_points)
+    mount_points.each do |_, target|
+      mount_points.each do |_, higher|
+        next if target == higher
+        # We're using +#mount_point+ on the tasks to use the normalized mount point
+        # names.
+        # If the higher mount point is found at the start of the target mount point.
+        # This will match /dev for /dev/shm, but not the reverse.
+        if target.mount_point.index(higher.mount_point) == 0 then
+          $logger.debug("#{target.mount_point} is under #{higher.mount_point}")
+          target.add_dependency(:Mount, higher.mount_point)
+        end
+      end
+    end
+  end
+
+  # Creates tasks to mount all special mount points.
+  # Dependencies on other mount points is being handled here.
+  def self.create_special_mount_points()
+    # Create all mount points.
+    mount_points = Configuration["nixos"]["boot"]["specialFileSystems"].map do |mount_point, config|
+      args = []
+      args << config["device"] if config["device"]
+      args << mount_point
+
+      task = Tasks::Mount.new(
+        *args,
+        type: config["fsType"],
+        options: config["options"],
+      )
+
+      [mount_point, task]
+    end.to_h
+    auto_depend_mount_points(mount_points)
+  end
+
+  # Creates tasks to mount all mount points required for boot.
+  # Dependencies on other mount points is being handled here.
+  def self.create_boot_mount_points()
+    mount_points = Configuration["bootFileSystems"].map do |mount_point, config|
+      mount_point = File.join(Tasks::SwitchRoot::SYSTEM_MOUNT_POINT, mount_point)
+
+      device = 
+        if config["label"]
+          "/dev/disk/by-label/#{config["label"]}"
+        else
+          config["device"]
+        end
+
+      options = config["options"].reject { |str| str.match(/^x-nixos\./) }
+
+      task = Tasks::Mount.new(
+        device,
+        mount_point,
+        type: config["fsType"],
+        options: options,
+      )
+
+      # TODO: Handle failures gracefully
+      # init_fail FFFF00 root_mount_failure "Could not mount root filesystem"
```

Comment:
#### TODO
 * [ ] Implement mount failure

Reply:
This will be moved to a new task, to be done outside this PR.

#### TODO

 * [ ] Implement dependency failures

Right now, all dependencies are expected to resolve at some point. This means that any failure in a dependency to be fulfilled means it will be stuck in an infinite loop.

We might need to add timed delays and tries delays for dependencies.

Then, we need to transmit the error for a dependency to resolve to the tasks that depend on it. In turn, that function will be able to handle it.